components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/authenticators/AbstractOAuthAuthenticator.java

```diff
@@ -0,0 +1,140 @@
+    /**
+     * @return rest API invalid token cache
+     */
+     public Cache getRESTAPIInvalidTokenCache() {
+        return CacheProvider.getRESTAPIInvalidTokenCache();
+    }
+
+    /**
+     * @param message   CXF message to be validate
+     * @param tokenInfo Token information associated with incoming request
+     * @return return true if we found matching scope in resource and token information
+     * else false(means scope validation failed).
+     */
+    @MethodStats
+    public boolean validateScopes(Message message, OAuthTokenInfo tokenInfo) {
+        String basePath = (String) message.get(Message.BASE_PATH);
+        // path is obtained from Message.REQUEST_URI instead of Message.PATH_INFO, as Message.PATH_INFO contains
+        // decoded values of request parameters
+        String path = (String) message.get(Message.REQUEST_URI);
+        String verb = (String) message.get(Message.HTTP_REQUEST_METHOD);
+        String resource = path.substring(basePath.length() - 1);
+        String[] scopes = tokenInfo.getScopes();
+
+        String version = (String) message.get(RestApiConstants.API_VERSION);
+
+        //get all the URI templates of the REST API from the base path
+        Set<URITemplate> uriTemplates = RestApiUtil.getURITemplatesForBasePath(basePath + version);
+        if (uriTemplates.isEmpty()) {
+            if (log.isDebugEnabled()) {
+                log.debug("No matching scopes found for request with path: " + basePath
+                        + ". Skipping scope validation.");
+            }
+            return true;
+        }
+
+        for (Object template : uriTemplates.toArray()) {
+            org.wso2.uri.template.URITemplate templateToValidate = null;
+            Map<String, String> var = new HashMap<String, String>();
+            //check scopes with what we have
+            String templateString = ((URITemplate) template).getUriTemplate();
+            try {
+                templateToValidate = new org.wso2.uri.template.URITemplate(templateString);
+            } catch (URITemplateException e) {
+                log.error("Error while creating URI Template object to validate request. Template pattern: " +
+                        templateString, e);
+            }
+            if (templateToValidate != null && templateToValidate.matches(resource, var) && scopes != null
+                    && verb != null && verb.equalsIgnoreCase(((URITemplate) template).getHTTPVerb())) {
+                for (String scope : scopes) {
+                    Scope scp = ((URITemplate) template).getScope();
+                    if (scp != null) {
+                        if (scope.equalsIgnoreCase(scp.getKey())) {
+                            //we found scopes matches
+                            if (log.isDebugEnabled()) {
+                                log.debug("Scope validation successful for access token: " +
+                                        tokenInfo.getAccessToken() + " with scope: " + scp.getKey() +
```

Comment:
Shall we use maked token method in this class also and add it in all missing places?

Reply:
To generate a masked token we need to get the JWT header from the singedJWT object. since it take high cpu cost, is it ok to send the masked token itself to this method and use ?

Reply:
This is the common scope validation path and not specific to the JWT authentication (can have both JWT and opaque token). Can't we get it from Message (adding the masked token in a property)?